### PR TITLE
HDDS-5165. OM DB checkpoint servlet not accessible in a secure cluster

### PR DESCRIPTION
**Pull request by vivekratnavel**
## What changes were proposed in this pull request?

Update OM DB checkpoint servlet to enforce access restrictions only when both ACL and SPNGEO auth are enabled. Also, updated the rules to allow recon by default.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5165

## How was this patch tested?

Unit tests and tested in a real cluster.

### CODE REVIEW

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java [hunk 1]
```diff
@@ -83,114 +89,184 @@
...
+        getNumCheckpoints() > initialCheckpointCount);
+  }
+
+  @Test
+  public void testSpnegoEnabled() throws Exception {
+    cluster.shutdown();
+    conf.setBoolean(OZONE_ACL_ENABLED, true);
+    conf.set(OZONE_ADMINISTRATORS, "");
+    conf.set(OZONE_OM_HTTP_AUTH_TYPE, "kerberos");
+    conf.set(OZONE_RECON_KERBEROS_PRINCIPAL_KEY, "recon/host1@REALM");
```
Comment (avijayanhwx) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:202]: There seems to be some common code between the 2 tests. Can they be moved to common method?

**Review by avijayanhwx — COMMENTED** [review]
Approach looks OK to me. 1 comment inline. Can we fix the findbugs issue?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java [hunk 2]
```diff
@@ -79,14 +86,27 @@ public void initialize(DBStore store, DBCheckpointMetrics metrics,
...
     this.ozAdmins = ozoneAdmins;
+    this.isSpnegoEnabled = isSpnegoEnabled;
+    ReconConfig reconConfig = conf.getObject(ReconConfig.class);
+    String reconPrincipal = reconConfig.getKerberosPrincipal();
+    if (!reconPrincipal.isEmpty()) {
+      UserGroupInformation ugi =
+          UserGroupInformation.createRemoteUser(reconPrincipal);
+      reconSPN = ugi.getShortUserName();
+    } else {
+      reconSPN = StringUtils.EMPTY;
```
Comment (xiaoyuyao) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java:97]: Do we have guarantee that SCM configuration always have recon related configuration like recon SPN?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java [hunk 3]
```diff
@@ -64,10 +67,14 @@
...
 
   private boolean aclEnabled;
+  private boolean isSpnegoEnabled;
   private Collection<String> ozAdmins;
+  private String reconSPN;
 
+  @SuppressWarnings("checkstyle:hiddenfield")
   public void initialize(DBStore store, DBCheckpointMetrics metrics,
-      boolean omAclEnabled, Collection<String> ozoneAdmins)
+                         boolean omAclEnabled, Collection<String> ozoneAdmins,
```
Comment (xiaoyuyao) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java:76]: Can we refactor the initialize() with just a Collection<String> allowedUsers? and let the callers to combine ozoneAdmin and recon SPN without additional checkstyle warning suppress?

**Review by xiaoyuyao — COMMENTED** [review]
Thanks @vivekratnavel for working on this. The PR looks good to me overall. A few minor suggestions added inline...

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java [hunk 4]
```diff
@@ -83,114 +93,171 @@
...
+    Assert.assertTrue(
+        omMetrics.getDBCheckpointMetrics().
+            getLastCheckpointStreamingTimeTaken() > 0);
+    Assert.assertTrue(omMetrics.getDBCheckpointMetrics().
+        getNumCheckpoints() > initialCheckpointCount);
+  }
+
+  @Test
+  public void testSpnegoEnabled() throws Exception {
+    cluster.shutdown();
```
Comment (xiaoyuyao) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:212]: Should we have cluster.shutdown() in testDoGet() to ensure we don't leak cluster instances running if the test execution order is changed?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java [hunk 5]
```diff
@@ -83,114 +93,171 @@
...
+    when(userPrincipalMock.getName()).thenReturn("recon/localhost@REALM");
+
+    when(requestMock.getUserPrincipal()).thenReturn(userPrincipalMock);
+    when(responseMock.getOutputStream()).thenReturn(servletOutputStream);
+
+    omDbCheckpointServletMock.doGet(requestMock, responseMock);
 
+    // Recon user should be able to access the servlet and download the
+    // snapshot
+    Assert.assertTrue(tempFile.length() > 0);
```
Comment (xiaoyuyao) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:260]: the test cluster instance needs to be shutdown at the end of the test.

**Review by xiaoyuyao — COMMENTED** [review]
Thanks for the update. I have one comment on the unit test cluster shutdown. Otherwise, LGTM.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java (hunk 4 above)
Comment (vivekratnavel) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:212]: Yes, it is handled in the `@AfterClass` method - https://github.com/apache/ozone/pull/2200/files#diff-816030b8fda0e307628b659606cf1d37b2117996c83caab10513b1956f72bc13R128

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java (hunk 5 above)
Comment (vivekratnavel) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:260]: Yes, it is handled in the `@AfterClass` method - https://github.com/apache/ozone/pull/2200/files#diff-816030b8fda0e307628b659606cf1d37b2117996c83caab10513b1956f72bc13R128

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java (hunk 4 above)
Comment (xiaoyuyao) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:212]: That should be sufficient for the current two test cases. My suggestion is to move the shutdown with @After for each test case. This way, new test cases added won't need to put cluster.shutdown() any more.

**Review by xiaoyuyao — APPROVED** [review]
+1 pending CI.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java (hunk 4 above)
Comment (vivekratnavel) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java]: Thanks @xiaoyuyao! I have updated the latest patch to fix this.

**Comment by vivekratnavel** [comment]
Thank you @xiaoyuyao and @avijayanhwx for the reviews.

**Pull request merged**